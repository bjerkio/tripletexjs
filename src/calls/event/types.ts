import { DefaultTripletexInputs } from '../../types';

export type ListSubscriptionsInput = DefaultTripletexInputs;

export interface CreateSubscriptionInput {
  /**
   * Event name (from /v2/event) you wish to subscribe to.
   *
   * Form should be: subject.verb.
   */
  event: string;

  /**
   * @example https://username:password@myintegration.example/tripletexCallback
   */
  targetUrl: string;

  /**
   * The fields in the object delivered
   * with the notification callback, nested as in other API calls.
   */
  fields?: string[];

  /**
   * Custom authentication header name
   */
  authHeaderName?: string;

  /**
   * Custom authentication header value (write only)
   */
  authHeaderValue?: string;
}
