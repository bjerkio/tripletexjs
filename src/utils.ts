import * as rt from 'runtypes';

export function pickFromObject<
  T extends Record<string, unknown>,
  K extends keyof T,
>(subject: T, ...keys: K[]): Pick<T, K> {
  const pairs = keys
    .map(key => [key, subject[key]])
    .filter(([, val]) => val !== undefined)
    .map(([key, val]) => [key, val]);
  return Object.fromEntries(pairs);
}

export function withRuntype<T>(validator: rt.Runtype<T>) {
  return (data: unknown) => {
    return validator.check(data);
  };
}

export function multipleValuesEnvelope<T>(type: rt.Runtype<T>) {
  return rt
    .Record({
      from: rt.Number,
      count: rt.Number,
      versionDigest: rt.String.nullable().optional(),
      values: rt.Array(type).asReadonly(),
    })
    .asReadonly();
}

export function singleValueEnvelope<T>(type: rt.Runtype<T>) {
  return rt
    .Record({
      value: type,
    })
    .asReadonly();
}
