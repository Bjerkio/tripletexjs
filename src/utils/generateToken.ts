import { OpenAPI } from '../generated/core/OpenAPI';
import { TokenSessionService } from '../generated/services/TokenSessionService';
import { invariant } from './invariant';

/**
 * generateToken helps creating a token. It sets token for all requests.
 * 
 * @param consumerToken Token of the API consumer
 * @param employeeToken The employees token
 * @param expirationDate Expiration date for the combined token
 * @example
 * ```typescript
 * const expDate = new Date();
 * expDate.setDate(expDate.getDate() + 1);
 * await generateToken({
 *   consumerToken: '000-000-000',
 *   employeeToken: '100-100-100',
 *   expirationDate: expDate,
 * });
 * ```
 */
export async function generateToken({
  expirationDate,
  ...tokenSessionArgs
}: {
  consumerToken: string;
  employeeToken: string;
  expirationDate: Date;
}) {
  if (OpenAPI.TOKEN) {
    return false;
  }

  const expDate = expirationDate.toISOString().split('T')[0];

  const session = await TokenSessionService.tokenSessionCreateCreate({
    expirationDate: expDate,
    ...tokenSessionArgs,
  });

  invariant(session.value.token, 'Not able to get token');

  OpenAPI.TOKEN = session.value;
}

/**
 * 0 or blank means the company of the employee.
 * Any other value means accountant clients. Use /company/>withLoginAccess to get a list of those.
 * @param username defaults to '0'
 */
export async function setUsername(username: string | (() => Promise<string>)) {
  OpenAPI.USERNAME = username;
}