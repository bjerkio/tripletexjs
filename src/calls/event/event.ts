import { serializeQuery, withRuntype } from '../../utils';
import { TripletexBase } from '../base';
import {
  createSubscriptionResponseRt,
  getSubscriptionResponseRt,
  listSubscriptionResponseRt,
} from './models/subscription';
import { CreateSubscriptionInput, ListSubscriptionsInput } from './types';
export * from './models/subscription';

export class TripletexEvent extends TripletexBase {
  getSubscription(id: number) {
    const call = this.authenticatedCall() //
      .args<{ id: number }>()
      .path(({ id }) => `/v2/event/subscription/${id}`)
      .method('get')
      .parseJson(withRuntype(getSubscriptionResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ id, sessionToken }));
  }

  deleteSubscription(id: number) {
    const call = this.authenticatedCall() //
      .args<{ id: number }>()
      .path(({ id }) => `/v2/event/subscription/${id}`)
      .method('delete')
      .build();

    return this.performRequest(sessionToken => call({ id, sessionToken }));
  }

  listSubscriptions(input?: ListSubscriptionsInput) {
    const call = this.authenticatedCall() //
      .args<{
        input?: ListSubscriptionsInput;
      }>()
      .path('/v2/event/subscription')
      .query(args => (args.input ? serializeQuery(args.input) : {}))
      .method('get')
      .parseJson(withRuntype(listSubscriptionResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }

  createSubscription(input: CreateSubscriptionInput) {
    const call = this.authenticatedCall() //
      .args<{
        input: CreateSubscriptionInput;
      }>()
      .path('/v2/event/subscription')
      .body(({ input }) => {
        return {
          ...input,
        };
      })
      .method('post')
      .parseJson(withRuntype(createSubscriptionResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }
}
