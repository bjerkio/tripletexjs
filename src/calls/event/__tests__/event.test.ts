import nock from 'nock';
import { invariant } from 'ts-invariant';
import { parseRuntypeValidationError } from '../../../utils';
import { TripletexEvent } from '../event';

const baseUrl = process.env.BASE_URL || 'https://api.tripletex.io/';
const sessionToken = process.env.SESSION_TOKEN || 'a-session-token';

describe('activity class', () => {
  let client: TripletexEvent;

  beforeEach(() => {
    client = new TripletexEvent({ baseUrl }, sessionToken);
  });

  it('should create a subscription', async () => {
    nock('https://api.tripletex.io:443', { encodedQueryParams: true })
      .post('/v2/event/subscription', {
        event: 'invoice.charged',
        targetUrl: 'http://example.com',
      })
      .reply(201, {
        value: {
          id: 48620,
          version: 0,
          url: 'api.tripletex.io/v2/event/subscription/48620',
          event: 'invoice.charged',
          targetUrl: 'http://example.com',
          fields: null,
          status: 'ACTIVE',
          authHeaderName: null,
        },
      });
    const subscription = await client.createSubscription({
      event: 'invoice.charged',
      targetUrl: 'http://example.com',
    });

    parseRuntypeValidationError(subscription.error);
    invariant(subscription.success);

    expect(subscription.body.value).toMatchSnapshot();
  });

  it('should get a subscription', async () => {
    nock('https://api.tripletex.io:443', { encodedQueryParams: true })
      .get('/v2/event/subscription/48620')
      .reply(200, {
        value: {
          id: 48620,
          version: 0,
          url: 'api.tripletex.io/v2/event/subscription/48620',
          event: 'invoice.charged',
          targetUrl: 'http://example.com',
          fields: null,
          status: 'ACTIVE',
          authHeaderName: null,
        },
      });
    const subscription = await client.getSubscription(48620);

    parseRuntypeValidationError(subscription.error);
    invariant(subscription.success);

    expect(subscription.body.value).toMatchSnapshot();
  });

  it('should delete a subscription', async () => {
    nock('https://api.tripletex.io:443', { encodedQueryParams: true })
      .delete('/v2/event/subscription/48619')
      .reply(204, '');
    const subscription = await client.deleteSubscription(48619);

    parseRuntypeValidationError(subscription.error);

    expect(subscription.success).toBeTruthy();
  });

  it('should list subscriptions', async () => {
    nock('https://api.tripletex.io:443', { encodedQueryParams: true })
      .get('/v2/event/subscription')
      .reply(200, {
        fullResultSize: 1,
        from: 0,
        count: 1,
        versionDigest: null,
        values: [
          {
            id: 48620,
            version: 0,
            url: 'api.tripletex.io/v2/event/subscription/48620',
            event: 'invoice.charged',
            targetUrl: 'http://example.com',
            fields: null,
            status: 'ACTIVE',
            authHeaderName: null,
          },
        ],
      });
    const subscription = await client.listSubscriptions();

    parseRuntypeValidationError(subscription.error);
    invariant(subscription.success);

    expect(subscription.body.values).toMatchSnapshot();
  });

  it('should create a subscription with fields', async () => {
    nock('https://api.tripletex.io:443', { encodedQueryParams: true })
      .post('/v2/event/subscription', {
        event: 'project.create',
        targetUrl: 'http://example.com',
        fields: 'id,name,number,isClosed',
      })
      .reply(201, {
        value: {
          id: 48638,
          version: 0,
          url: 'api.tripletex.io/v2/event/subscription/48638',
          event: 'project.create',
          targetUrl: 'http://example.com',
          fields: 'id,name,number,isClosed',
          status: 'ACTIVE',
          authHeaderName: null,
        },
      });
    const subscription = await client.createSubscription({
      event: 'project.create',
      targetUrl: 'http://example.com',
      fields: ['id', 'name', 'number', 'isClosed'],
    });

    parseRuntypeValidationError(subscription.error);
    invariant(subscription.success);

    expect(subscription.body.value).toMatchSnapshot();
  });
});
