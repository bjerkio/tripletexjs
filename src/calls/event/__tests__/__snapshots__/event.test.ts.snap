// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activity class should create a subscription 1`] = `
Object {
  "authHeaderName": null,
  "event": "invoice.charged",
  "fields": null,
  "id": 48620,
  "status": "ACTIVE",
  "targetUrl": "http://example.com",
  "url": "api.tripletex.io/v2/event/subscription/48620",
  "version": 0,
}
`;

exports[`activity class should get a subscription 1`] = `
Object {
  "authHeaderName": null,
  "event": "invoice.charged",
  "fields": null,
  "id": 48620,
  "status": "ACTIVE",
  "targetUrl": "http://example.com",
  "url": "api.tripletex.io/v2/event/subscription/48620",
  "version": 0,
}
`;

exports[`activity class should list subscriptions 1`] = `
Array [
  Object {
    "authHeaderName": null,
    "event": "invoice.charged",
    "fields": null,
    "id": 48620,
    "status": "ACTIVE",
    "targetUrl": "http://example.com",
    "url": "api.tripletex.io/v2/event/subscription/48620",
    "version": 0,
  },
]
`;
